File: src/main/java/com/tkr/desk/MarginAggregator.java
package com.tkr.desk;

import com.tkr.types.WireTypes;
import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Margin aggregator with SPAN scanning risk. Ported from legacy-cpp margin_aggregator.cc.
 */
public final class MarginAggregator {

    public static final class MarginLine {
        public int accountId;
        public int symbolId;
        public int qtyMilli;
        public long notionalCents;
        public long initialMarginCents;
        public long maintenanceMarginCents;
    }

    public static final class MarginAggregatorConfig {
        public int deskId;
        public int marginRateBp;

        public MarginAggregatorConfig(int deskId, int marginRateBp) {
            this.deskId = deskId;
            this.marginRateBp = marginRateBp;
        }
    }

    public static final class MarginAggregatorResult {
        public Status status = Status.OK;
        public List<MarginLine> lines = new ArrayList<>();
        public List<MarginCheckResult> accountChecks = new ArrayList<>();
        public long totalInitialMarginCents;
        public int breachCount;
    }

    /** Legacy wrapper for DeskRegistry. */
    public static final class AccountMarginState {
        public int accountId;
        public long requiredMarginCents;
        public long availableMarginCents;
        public boolean passed;
    }

    public static final class MarginBatchResult {
        public Status status = Status.OK;
        public List<AccountMarginState> accounts = new ArrayList<>();
        public boolean allPassed = true;
    }

    private static final class AccountRollup {
        int accountId;
        long initialMarginCents;
        long maintenanceMarginCents;
        long notionalCents;
        int lineCount;
    }

    private static final class SymbolExposure {
        int symbolId;
        long netQtyMilli;
        long grossNotionalCents;
        long spanScanningRiskCents;
    }

    private static final class SpanScenario {
        int priceShiftBp;
        int volShiftBp;
        long pnlCents;

        SpanScenario(int priceShiftBp, int volShiftBp, long pnlCents) {
            this.priceShiftBp = priceShiftBp;
            this.volShiftBp = volShiftBp;
            this.pnlCents = pnlCents;
        }
    }

    private static final long DEFAULT_AVAILABLE_CENTS = 1_000_000_000L;
    private static final int MAINT_RATIO_BP = 7500;
    private static final int DEFAULT_MARGIN_RATE_BP = 1500;

    private final MarginAggregatorConfig config;
    private final Map<Integer, Long> accountAvailable = new HashMap<>();

    public MarginAggregator() {
        this(new MarginAggregatorConfig(0, DEFAULT_MARGIN_RATE_BP));
    }

    public MarginAggregator(MarginAggregatorConfig config) {
        this.config = config != null ? config : new MarginAggregatorConfig(0, DEFAULT_MARGIN_RATE_BP);
    }

    public void setAvailableMarginCents(int accountId, long cents) {
        accountAvailable.put(accountId, cents);
    }

    public void clearAccountMargins() {
        accountAvailable.clear();
    }

    public static long computeNotionalCents(int qtyMilli, int priceTick) {
        return (long) qtyMilli * (long) priceTick / 1000L;
    }

    public static long computeInitialMargin(long notionalCents, int rateBp) {
        if (notionalCents <= 0) return 0;
        return notionalCents * rateBp / 10000L;
    }

    public static long computeMaintenanceMargin(long initialMargin, int maintRatioBp) {
        if (initialMargin <= 0) return 0;
        return initialMargin * maintRatioBp / 10000L;
    }

    private long lookupAvailableMargin(int accountId) {
        return accountAvailable.getOrDefault(accountId, DEFAULT_AVAILABLE_CENTS);
    }

    private static long applySpanPriceScan(long notionalCents, int shiftBp) {
        return notionalCents * shiftBp / 10000L;
    }

    private static long computeSpanScanningRisk(SymbolExposure exposure) {
        SpanScenario[] scenarios = {
            new SpanScenario(500, 0, applySpanPriceScan(exposure.grossNotionalCents, 500)),
            new SpanScenario(-500, 0, applySpanPriceScan(exposure.grossNotionalCents, -500)),
            new SpanScenario(300, 200, applySpanPriceScan(exposure.grossNotionalCents, 300)),
            new SpanScenario(-300, 200, applySpanPriceScan(exposure.grossNotionalCents, -300)),
            new SpanScenario(0, 500, applySpanPriceScan(exposure.grossNotionalCents, 250))
        };
        long worstLoss = 0;
        for (SpanScenario sc : scenarios) {
            if (sc.pnlCents < worstLoss) worstLoss = sc.pnlCents;
        }
        return -worstLoss;
    }

    public Status buildMarginLines(BatchWireFrame frame, List<MarginLine> lines) {
        if (lines == null) return Status.BOUNDS_ERROR;
        lines.clear();
        Map<Integer, SymbolExposure> exposures = new HashMap<>();
        for (WireBatchRecord rec : frame.records) {
            MarginLine line = new MarginLine();
            line.accountId = rec.accountId;
            line.symbolId = rec.symbolId;
            line.qtyMilli = rec.qtyMilli;
            line.notionalCents = computeNotionalCents(rec.qtyMilli, rec.priceTick);
            SymbolExposure exp = exposures.computeIfAbsent(rec.symbolId, id -> {
                SymbolExposure s = new SymbolExposure();
                s.symbolId = id;
                return s;
            });
            exp.netQtyMilli += rec.qtyMilli;
            exp.grossNotionalCents += line.notionalCents;
            line.initialMarginCents = computeInitialMargin(line.notionalCents, config.marginRateBp);
            line.maintenanceMarginCents = computeMaintenanceMargin(line.initialMarginCents, MAINT_RATIO_BP);
            lines.add(line);
        }
        for (MarginLine line : lines) {
            SymbolExposure exp = exposures.get(line.symbolId);
            if (exp == null) continue;
            long spanRisk = computeSpanScanningRisk(exp);
            if (spanRisk > line.initialMarginCents) {
                line.initialMarginCents = spanRisk;
                line.maintenanceMarginCents = computeMaintenanceMargin(spanRisk, MAINT_RATIO_BP);
            }
        }
        return Status.OK;
    }

    public Status rollUpByAccount(List<MarginLine> lines, List<MarginCheckResult> checks) {
        if (checks == null) return Status.BOUNDS_ERROR;
        Map<Integer, AccountRollup> rollup = new HashMap<>();
        for (MarginLine line : lines) {
            AccountRollup acc = rollup.computeIfAbsent(line.accountId, id -> {
                AccountRollup a = new AccountRollup();
                a.accountId = id;
                return a;
            });
            acc.initialMarginCents += line.initialMarginCents;
            acc.maintenanceMarginCents += line.maintenanceMarginCents;
            acc.notionalCents += line.notionalCents;
            acc.lineCount++;
        }
        checks.clear();
        for (AccountRollup acc : rollup.values()) {
            MarginCheckResult check = new MarginCheckResult();
            check.accountId = acc.accountId;
            check.requiredMarginCents = acc.initialMarginCents;
            check.availableMarginCents = lookupAvailableMargin(acc.accountId);
            check.passed = check.availableMarginCents >= check.requiredMarginCents;
            checks.add(check);
        }
        checks.sort(Comparator.comparingInt(c -> c.accountId));
        return Status.OK;
    }

    public MarginAggregatorResult aggregate(BatchWireFrame frame) {
        MarginAggregatorResult result = new MarginAggregatorResult();
        result.status = Status.OK;
        if (frame == null || frame.records.isEmpty()) return result;
        Status buildSt = buildMarginLines(frame, result.lines);
        if (buildSt != Status.OK) {
            result.status = buildSt;
            return result;
        }
        for (MarginLine line : result.lines) result.totalInitialMarginCents += line.initialMarginCents;
        Status rollSt = rollUpByAccount(result.lines, result.accountChecks);
        if (rollSt != Status.OK) {
            result.status = rollSt;
            return result;
        }
        for (MarginCheckResult check : result.accountChecks) {
            if (!check.passed) result.breachCount++;
        }
        if (result.breachCount > 0 && (frame.header.flags & WireTypes.BATCH_FLAG_MARGIN_CHECK) != 0) {
            result.status = Status.MARGIN_BREACH;
        }
        return result;
    }

    public long computeRequiredMarginCents(int qtyMilli, int priceTick) {
        return computeInitialMargin(computeNotionalCents(qtyMilli, priceTick), config.marginRateBp);
    }

    /** Legacy API. */
    public MarginBatchResult aggregateBatch(BatchWireFrame frame) {
        MarginAggregatorResult agg = aggregate(frame);
        MarginBatchResult result = new MarginBatchResult();
        result.status = agg.status;
        result.allPassed = agg.breachCount == 0;
        for (MarginCheckResult check : agg.accountChecks) {
            AccountMarginState state = new AccountMarginState();
            state.accountId = check.accountId;
            state.requiredMarginCents = check.requiredMarginCents;
            state.availableMarginCents = check.availableMarginCents;
            state.passed = check.passed;
            result.accounts.add(state);
            if (!state.passed) result.allPassed = false;
        }
        if (!result.allPassed && result.status == Status.OK) result.status = Status.MARGIN_BREACH;
        return result;
    }

    public MarginCheckResult toCheckResult(AccountMarginState state) {
        MarginCheckResult r = new MarginCheckResult();
        r.accountId = state.accountId;
        r.requiredMarginCents = state.requiredMarginCents;
        r.availableMarginCents = state.availableMarginCents;
        r.passed = state.passed;
        return r;
    }

    public long totalRequiredCents(MarginBatchResult result) {
        long sum = 0;
        for (AccountMarginState s : result.accounts) sum += s.requiredMarginCents;
        return sum;
    }

    public AccountMarginState worstAccount(MarginBatchResult result) {
        AccountMarginState worst = null;
        for (AccountMarginState s : result.accounts) {
            if (!s.passed && (worst == null || s.requiredMarginCents > worst.requiredMarginCents)) worst = s;
        }
        return worst;
    }

    public long totalMaintenanceCents(MarginAggregatorResult result) {
        long sum = 0;
        for (MarginLine line : result.lines) sum += line.maintenanceMarginCents;
        return sum;
    }

    public List<MarginLine> linesForAccount(MarginAggregatorResult result, int accountId) {
        List<MarginLine> filtered = new ArrayList<>();
        for (MarginLine line : result.lines) {
            if (line.accountId == accountId) filtered.add(line);
        }
        return filtered;
    }

    public long spanRiskForSymbol(BatchWireFrame frame, int symbolId) {
        SymbolExposure exp = new SymbolExposure();
        exp.symbolId = symbolId;
        for (WireBatchRecord rec : frame.records) {
            if (rec.symbolId != symbolId) continue;
            exp.netQtyMilli += rec.qtyMilli;
            exp.grossNotionalCents += computeNotionalCents(rec.qtyMilli, rec.priceTick);
        }
        return computeSpanScanningRisk(exp);
    }

    public boolean passesMarginCheck(int accountId, long requiredCents) {
        return lookupAvailableMargin(accountId) >= requiredCents;
    }

    public Map<Integer, Long> snapshotAvailable() {
        return new HashMap<>(accountAvailable);
    }

    public int marginRateBp() {
        return config.marginRateBp;
    }

    public String summarizeResult(MarginAggregatorResult result) {
        return "lines=" + result.lines.size() + " breaches=" + result.breachCount
                + " totalInitial=" + result.totalInitialMarginCents;
    }

    public MarginCheckResult worstBreach(MarginAggregatorResult result) {
        MarginCheckResult worst = null;
        for (MarginCheckResult check : result.accountChecks) {
            if (!check.passed && (worst == null || check.requiredMarginCents > worst.requiredMarginCents)) {
                worst = check;
            }
        }
        return worst;
    }

    public long headroomCents(int accountId, MarginAggregatorResult result) {
        for (MarginCheckResult check : result.accountChecks) {
            if (check.accountId == accountId) {
                return check.availableMarginCents - check.requiredMarginCents;
            }
        }
        return lookupAvailableMargin(accountId);
    }

    public int breachAccountCount(MarginAggregatorResult result) {
        return result.breachCount;
    }
}

File: legacy-cpp/src/desk/margin_aggregator.cc
#include "tkr/desk/margin_aggregator.h"

#include <algorithm>
#include <unordered_map>
#include <vector>

namespace tkr {
namespace desk {
namespace {

struct AccountRollup {
  std::uint32_t account_id;
  std::int64_t initial_margin_cents;
  std::int64_t maintenance_margin_cents;
  std::int64_t notional_cents;
  std::uint32_t line_count;
};

struct SymbolExposure {
  std::uint32_t symbol_id;
  std::int64_t net_qty_milli;
  std::int64_t gross_notional_cents;
  std::int64_t span_scanning_risk_cents;
};

struct SpanScenario {
  std::int32_t price_shift_bp;
  std::int32_t vol_shift_bp;
  std::int64_t pnl_cents;
};

std::int64_t ApplySpanPriceScan(std::int64_t notional_cents,
                                std::int32_t shift_bp) {
  return (notional_cents * shift_bp) / 10000;
}

std::int64_t ComputeSpanScanningRisk(const SymbolExposure& exposure) {
  const SpanScenario scenarios[] = {
      {500, 0, ApplySpanPriceScan(exposure.gross_notional_cents, 500)},
      {-500, 0, ApplySpanPriceScan(exposure.gross_notional_cents, -500)},
      {300, 200, ApplySpanPriceScan(exposure.gross_notional_cents, 300)},
      {-300, 200, ApplySpanPriceScan(exposure.gross_notional_cents, -300)},
      {0, 500, ApplySpanPriceScan(exposure.gross_notional_cents, 250)},
  };

  std::int64_t worst_loss = 0;
  for (const SpanScenario& sc : scenarios) {
    if (sc.pnl_cents < worst_loss) {
      worst_loss = sc.pnl_cents;
    }
  }
  return -worst_loss;
}

}  // namespace

MarginAggregator::MarginAggregator(MarginAggregatorConfig config)
    : config_(config) {}

void MarginAggregator::SetAccountAvailableMargin(std::uint32_t account_id,
                                                 std::int64_t available_cents) {
  account_available_[account_id] = available_cents;
}

void MarginAggregator::ClearAccountMargins() { account_available_.clear(); }

std::int64_t MarginAggregator::ComputeNotionalCents(std::uint32_t qty_milli,
                                                    std::uint32_t price_tick) {
  return (static_cast<std::int64_t>(qty_milli) *
          static_cast<std::int64_t>(price_tick)) /
         1000;
}

std::int64_t MarginAggregator::ComputeInitialMargin(std::int64_t notional_cents,
                                                    std::uint32_t rate_bp) {
  if (notional_cents <= 0) {
    return 0;
  }
  return (notional_cents * static_cast<std::int64_t>(rate_bp)) / 10000;
}

std::int64_t MarginAggregator::ComputeMaintenanceMargin(
    std::int64_t initial_margin, std::uint32_t maint_ratio_bp) {
  if (initial_margin <= 0) {
    return 0;
  }
  return (initial_margin * static_cast<std::int64_t>(maint_ratio_bp)) / 10000;
}

std::int64_t MarginAggregator::LookupAvailableMargin(
    std::uint32_t account_id) const {
  const auto it = account_available_.find(account_id);
  if (it == account_available_.end()) {
    return 1000000000;
  }
  return it->second;
}

Status MarginAggregator::BuildMarginLines(const BatchWireFrame& frame,
                                          std::vector<MarginLine>* lines) const {
  if (lines == nullptr) {
    return Status::kBoundsError;
  }

  lines->clear();
  lines->reserve(frame.records.size());

  std::unordered_map<std::uint32_t, SymbolExposure> exposures;

  for (const WireBatchRecord& rec : frame.records) {
    MarginLine line{};
    line.account_id = rec.account_id;
    line.symbol_id = rec.symbol_id;
    line.qty_milli = rec.qty_milli;
    line.notional_cents = ComputeNotionalCents(rec.qty_milli, rec.price_tick);

    SymbolExposure& exp = exposures[rec.symbol_id];
    exp.symbol_id = rec.symbol_id;
    exp.net_qty_milli += static_cast<std::int64_t>(rec.qty_milli);
    exp.gross_notional_cents += line.notional_cents;

    line.initial_margin_cents =
        ComputeInitialMargin(line.notional_cents, config_.margin_rate_bp);
    line.maintenance_margin_cents =
        ComputeMaintenanceMargin(line.initial_margin_cents, 7500);
    lines->push_back(line);
  }

  for (MarginLine& line : *lines) {
    const SymbolExposure& exp = exposures[line.symbol_id];
    const std::int64_t span_risk = ComputeSpanScanningRisk(exp);
    if (span_risk > line.initial_margin_cents) {
      line.initial_margin_cents = span_risk;
      line.maintenance_margin_cents =
          ComputeMaintenanceMargin(span_risk, 7500);
    }
  }

  return Status::kOk;
}

Status MarginAggregator::RollUpByAccount(
    const std::vector<MarginLine>& lines,
    std::vector<MarginCheckResult>* checks) const {
  if (checks == nullptr) {
    return Status::kBoundsError;
  }

  std::unordered_map<std::uint32_t, AccountRollup> rollup;

  for (const MarginLine& line : lines) {
    AccountRollup& acc = rollup[line.account_id];
    acc.account_id = line.account_id;
    acc.initial_margin_cents += line.initial_margin_cents;
    acc.maintenance_margin_cents += line.maintenance_margin_cents;
    acc.notional_cents += line.notional_cents;
    acc.line_count += 1;
  }

  checks->clear();
  checks->reserve(rollup.size());

  for (const auto& pair : rollup) {
    const AccountRollup& acc = pair.second;
    MarginCheckResult check{};
    check.account_id = acc.account_id;
    check.required_margin_cents = acc.initial_margin_cents;
    check.available_margin_cents = LookupAvailableMargin(acc.account_id);
    check.passed =
        check.available_margin_cents >= check.required_margin_cents;
    checks->push_back(check);
  }

  std::sort(checks->begin(), checks->end(),
            [](const MarginCheckResult& a, const MarginCheckResult& b) {
              return a.account_id < b.account_id;
            });

  return Status::kOk;
}

MarginAggregatorResult MarginAggregator::Aggregate(const BatchWireFrame& frame) {
  MarginAggregatorResult result{};
  result.status = Status::kOk;

  if (frame.records.empty()) {
    return result;
  }

  Status build_st = BuildMarginLines(frame, &result.lines);
  if (build_st != Status::kOk) {
    result.status = build_st;
    return result;
  }

  result.total_initial_margin_cents = 0;
  for (const MarginLine& line : result.lines) {
    result.total_initial_margin_cents += line.initial_margin_cents;
  }

  Status roll_st = RollUpByAccount(result.lines, &result.account_checks);
  if (roll_st != Status::kOk) {
    result.status = roll_st;
    return result;
  }

  for (const MarginCheckResult& check : result.account_checks) {
    if (!check.passed) {
      ++result.breach_count;
    }
  }

  if (result.breach_count > 0 &&
      (frame.header.flags & kBatchFlagMarginCheck) != 0) {
    result.status = Status::kMarginBreach;
  }

  return result;
}

}  // namespace desk
}  // namespace tkr
